### PR TITLE
Type link

### LINKED ISSUE #21: support linking between types in document

### LINKED ISSUE #22: show and link superclasses / traits in output

### LINKED ISSUE #31: support nested / path dependent type
```scala
class A{ class B }
val a1 = new A
val a2 = new A
val b: A#B = new a2.B (edited)
class A{ class B; def b: B }
```

### PR DESCRIPTION
**Pull request by kolowheel**
fixes #21

### CODE REVIEW

File: src/main/scala/DocGenerator.scala [hunk 1]
```diff
@@ -245,7 +251,45 @@ object LatexDocGenerator extends DocGenerator {
...
+    else
+      s"\\hyperlink{${link(e.id)}}{$t}"
+  }
+
+  def link(tpe: Seq[Tree]) = {
+    val termToTerm = "."
+    val termToType = "."
+    val typeToType = "#"
+    val typeToTerm = "_"
+    def separtor(s: Tree, i: Int): String = if (s != tpe.last) {
```
Comment (cvogt) [inline src/main/scala/DocGenerator.scala]: @clhodapp @xeno-by can Tree be compared this way in meta?

File: src/main/scala/DocGenerator.scala [hunk 2]
```diff
@@ -245,7 +251,45 @@ object LatexDocGenerator extends DocGenerator {
...
+      t
+    else
+      s"\\hyperlink{${link(e.id)}}{$t}"
+  }
+
+  def link(tpe: Seq[Tree]) = {
+    val termToTerm = "."
+    val termToType = "."
+    val typeToType = "#"
+    val typeToTerm = "_"
```
Comment (cvogt) [inline src/main/scala/DocGenerator.scala]: I think people usually use `#` for this, too

File: src/main/scala/DocGenerator.scala [hunk 3]
```diff
@@ -56,7 +60,7 @@ object LatexDocGenerator extends DocGenerator {
       case p: Pkg =>
         p +: loop(p)
     }.flatten
-    loop(root)
+    loop(root) :+ root
   }
```
Comment (cvogt) [inline src/main/scala/DocGenerator.scala]: I think an extractor method like this should go into a separate place where it can be re-used by multiple backends

File: src/main/scala/Main.scala [hunk 4]
```diff
@@ -2,6 +2,25 @@
...
+    println(LatexDocGenerator.dumpType(tpe))
+
+    val `[T]` = Type.Param(Nil, Type.Name("T",Seq()), Nil, Type.Bounds(None, None), Nil, Nil)
+    println(LatexDocGenerator.dumpType(`[T]` ))
+
+    //Type.Param(Nil, Type.Name("A"), Nil, Type.Bounds(None, None), List(Type.Apply(Type.Name("Ordered"), List(Type.Name("A")))), Nil)
+    val `[A <% Ordered[A]]` = Type.Param(Nil, Type.Name("A",Seq()), Nil, Type.Bounds(Some(Type.Name("Int",Seq())), None), List(Type.Apply(Type.Name("CustomClass",Seq(Term.Name("org",Seq()),Type.Name("CustomClass",Seq()))), List(Type.Name("A",Seq())))), Nil)
+    println(LatexDocGenerator.dumpType(`[A <% Ordered[A]]`))
+
+
```
Comment (cvogt) [inline src/main/scala/Main.scala]: why the duplication with the test?

File: src/test/scala/LatexDocGeneratorTest.scala [hunk 5]
```diff
@@ -4,7 +4,29 @@ import org.scalatest.FunSuite
...
+    val tpe = Type.Apply(
+      Type.Name("Seq", Seq(Term.Name("scala", Seq()), Term.Name("collection", Seq()), Type.Name("Seq", Seq()))),
+      Seq(intTpe))
+    assert(LatexDocGenerator.dumpType(tpe)=="\\hyperlink{scala.collection.Seq}{Seq}[\\hyperlink{scala.Int}{Int}]")
+
+    val `[T]` = Type.Param(Nil, Type.Name("T",Seq()), Nil, Type.Bounds(None, None), Nil, Nil)
+    assert(LatexDocGenerator.dumpType(`[T]` ) == "[T]")
+
+    //Type.Param(Nil, Type.Name("A"), Nil, Type.Bounds(None, None), List(Type.Apply(Type.Name("Ordered"), List(Type.Name("A")))), Nil)
+    val `[A <% Ordered[A]]` = Type.Param(Nil, Type.Name("A",Seq()), Nil, Type.Bounds(Some(Type.Name("Int",Seq())), None), List(Type.Apply(Type.Name("CustomClass",Seq(Term.Name("org",Seq()),Type.Name("CustomClass",Seq()))), List(Type.Name("A",Seq())))), Nil)
```
Comment (cvogt) [inline src/test/scala/LatexDocGeneratorTest.scala]: I love some indentation here, rather than one long line. Hard to follow

File: src/main/scala/DocGenerator.scala (hunk 1 above)
Comment (cvogt) [inline src/main/scala/DocGenerator.scala]: type in `separator`(missing an a)

File: src/main/scala/DocGenerator.scala [hunk 6]
```diff
@@ -67,130 +71,132 @@ object LatexDocGenerator extends DocGenerator {
...
+        val name = dumpType(e.name)
+        val tparams = dumpTypeParams(e.tparams)
+        val lo = e.typeBounds.lo.map(" >: " + dumpType(_)).getOrElse("")
+        val hi = e.typeBounds.hi.map(" <: " + dumpType(_)).getOrElse("")
+        val ctx = e.contextBounds.map(dumpType).map(": " + _).mkString(" ")
+        val view = e.viewBounds.map(dumpType).map("<% " + _).mkString(" ")
+        "[" + name + tparams + hi + lo + ctx + view + "]"
+      }
+    }
+  }
```
Comment (cvogt) [inline src/main/scala/DocGenerator.scala]: this is cool :)! I like this

**Comment by cvogt** [comment]
apart from the things I mentioned LGTM

File: src/main/scala/DocGenerator.scala (hunk 1 above)
Comment (clhodapp) [inline src/main/scala/DocGenerator.scala]: `can Tree be compared this way in meta?` I thought so... but getting confirmation from @xeno-by would be good

File: src/main/scala/DocGenerator.scala [hunk 7]
```diff
@@ -39,13 +38,18 @@ object PlainStringGenerator extends DocGenerator {
 object LatexDocGenerator extends DocGenerator {
 
 
+  def generateIndex(index: Index): String = {
+    "\\newpage\n" +
+      "Generated INDEX" + indexForObjects(index) + "\n" +
+      "Generated INDEX" + indexForMethods(index) + "\n" +
+      "Generated INDEX" + indexForClasses(index) + "\n" +
+      "Generated INDEX" + indexForTraits(index) + "\n"
+  }
```
Comment (clhodapp) [inline src/main/scala/DocGenerator.scala]: I'm not sure that this will produce good output. The X in INDEX and the return value of indexForX will be right up against each other with no spaces. Also, why aren't we using a triple-quoted String?

File: src/main/scala/DocGenerator.scala [hunk 8]
```diff
@@ -67,130 +71,132 @@ object LatexDocGenerator extends DocGenerator {
...
+      s"[$params]"
+    } else ""
+  }
+
+  def dumpType(tpe: Type): String = {
+
+    tpe match {
+      case e: Type.Name => hyperlink(e, Some(e.name))
+      case e: Type.Apply => dumpType(e.tpe) + "[" + e.args.map(dumpType).mkString(", ") + "]"
+      case e: Type.Compound => e.tpes.map(dumpType).mkString("with ")
```
Comment (clhodapp) [inline src/main/scala/DocGenerator.scala]: Will you end up with something like `Xwith Ywith Z`, since you do `mkString("with ")` ?

File: src/main/scala/newmodel/DocElement.scala [hunk 9]
```diff
@@ -20,33 +18,37 @@ sealed trait Defn extends Stat
 
 object Defn {
 
-  case class Object(name: Name,
+  case class Object(name: String,
```
Comment (clhodapp) [inline src/main/scala/newmodel/DocElement.scala]: Shouldn't this be a `Term.Name`?

File: src/main/scala/newmodel/DocElement.scala [hunk 10]
```diff
@@ -20,33 +18,37 @@ sealed trait Defn extends Stat
...
                     templ: Template,
                     comment: Comment,
                     mods: Seq[Mod],
-                    file: SourceFile) extends Defn
+                    file: SourceFile,
+                    id: Seq[Tree]) extends Defn
 
-  case class Trait(name: Name,
```
Comment (clhodapp) [inline src/main/scala/newmodel/DocElement.scala]: Shouldn't this be a `Type.Name`?

**Comment by clhodapp** [comment]
One general comment: I really don't like seeing String adding via `+`. I much prefer `++`. There's more overhead, I'll eat that in exchange for knowing that I'm not accidentally calling `toString` on things when I didn't mean to.

File: src/main/scala/DocGenerator.scala (hunk 1 above)
Comment (xeno-by) [inline src/main/scala/DocGenerator.scala]: What would you expect from this comparison?

File: src/main/scala/DocGenerator.scala (hunk 2 above)
Comment (xeno-by) [inline src/main/scala/DocGenerator.scala]: I don't think that this kind of path is expressible in Scala.

Comment (cvogt) [inline src/main/scala/DocGenerator.scala]: I have seen people use "Type#term" for that. It might be needed to refer to trait methods for example.

Comment (clhodapp) [inline src/main/scala/DocGenerator.scala]: It sort of feels like it makes sense for docs, even if it doesn't for code:
```scala
trait T {
  /** how many foos
   */
  val fooCount: Int
}
object T extends T {
  /** two foos
   */ 
 val fooCount = 2
}
```

You want docs for both the `fooCount` in the `object` and the `fooCount` in the `trait`... even if there isn't an actual expression of the path to the abstract `fooCount` in the `object` in Scala proper...

File: src/main/scala/DocGenerator.scala (hunk 1 above)
Comment (clhodapp) [inline src/main/scala/DocGenerator.scala]: Reference equality

Comment (clhodapp) [inline src/main/scala/DocGenerator.scala]: That said, why are we doing this this way?

File: src/main/scala/DocGenerator.scala (hunk 2 above)
Comment (clhodapp) [inline src/main/scala/DocGenerator.scala]: Edit there in the previous comment for you email users; I'm not sure if it sends you email notifications of edits.

File: src/main/scala/DocGenerator.scala (hunk 1 above)
Comment (xeno-by) [inline src/main/scala/DocGenerator.scala]: If you want reference equality, go for `eq`. If you want comparison that takes into account name resolution, go for `==` (but at the moment it will most likely not work).

Comment (cvogt) [inline src/main/scala/DocGenerator.scala]: actually guys, I just realize, these are not meta trees, but our doc trees. @clhodapp @kolowheel does != work for our trees?

**Comment by cvogt** [comment]
super type support LGTM
